### PR TITLE
Command to revoke privileges, opposite of grandfather command

### PR DESCRIPTION
**Pull request by gherceg**
## Summary

Need a way to revoke privileges. This takes what `cchq_prbac_grandfather_privs.py` and `ensure_grants` does and flips it to remove privileges. Lots of duplicate code, so if we would prefer me to consolidate this happy to do so. Wanted to at least make sure this was correct.
## Safety Assurance

- [x] Risk label is set correctly
- [x] All migrations are backwards compatible and won't block deploy
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change
- [x] If QA is part of the safety story, the "Awaiting QA" label is used
- [x] I have confidence that this PR will not introduce a regression for the reasons below

### Safety story

This is a management command. Will dry-run before running for real.
### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### CODE REVIEW

File: corehq/apps/accounting/utils/__init__.py [hunk 1]
```diff
@@ -303,6 +303,51 @@ def ensure_grants(grants_to_privs, dry_run=False, verbose=False, roles_by_slug=N
...
+        if grantee_slug not in roles_by_slug:
+            logger.info('grantee %s does not exist.', grantee_slug)
+            continue
+
+        for priv_slug in priv_slugs:
+            if priv_slug not in roles_by_slug:
+                logger.info('privilege %s does not exist.', priv_slug)
+                continue
+
+            if priv_slug not in granted[grantee_slug]:
```
Comment (gherceg) [inline corehq/apps/accounting/utils/__init__.py:332]: changed this line to first check if privilege is already revoked.

File: corehq/apps/accounting/utils/__init__.py [hunk 2]
```diff
@@ -303,6 +303,51 @@ def ensure_grants(grants_to_privs, dry_run=False, verbose=False, roles_by_slug=N
...
+                    logger.info('%Revoking privilege: %s => %s',
+                        dry_run_tag, grantee_slug, priv_slug)
+                if not dry_run:
+                    grants_to_revoke.append(Grant(
+                        from_role=roles_by_slug[grantee_slug],
+                        to_role=roles_by_slug[priv_slug]
+                    ))
+    if grants_to_revoke:
+        Role.get_cache().clear()
+        Grant.objects.bulk_delete(grants_to_revoke)
```
Comment (gherceg) [inline corehq/apps/accounting/utils/__init__.py:348]: changed this to delete grant

File: corehq/apps/accounting/utils/__init__.py [hunk 3]
```diff
@@ -303,6 +303,51 @@ def ensure_grants(grants_to_privs, dry_run=False, verbose=False, roles_by_slug=N
...
+            if priv_slug not in roles_by_slug:
+                logger.info('privilege %s does not exist.', priv_slug)
+                continue
+
+            if priv_slug not in granted[grantee_slug]:
+                if verbose or dry_run:
+                    logger.info('%sPrivilege already revoked: %s => %s',
+                        dry_run_tag, grantee_slug, priv_slug)
+            else:
+                granted[grantee_slug].discard(priv_slug)
```
Comment (gherceg) [inline corehq/apps/accounting/utils/__init__.py:337]: changed this to discard privilege from set if to be removed.

**Comment by gherceg** [comment]
fyi @orangejenny

**Review by biyeun — APPROVED** [review]
This looks good. Thank you!

**Comment by gherceg** [comment]
I've only tested dry running this command, but am confident the command does what I want. The tricky part is defining the process around using this. Looking at the process for creating privileges, we create a migration that ensures any new roles are created (`cchq_prbac_bootstrap`), and then ensures privileges are added to the appropriate roles (`cchq_prbac_grandfather_privs`). 

In this case, I plan to make the following changes on a separate PR:
- remove all references to the `LINKED_PROJECTS` privilege, replace with slug "linked_projects" where needed
- create a migration that calls this command for the slug "linked_projects" with the `delete_privs` flag set

That way, the `Role` with the `linked_projects` slug will no longer exist after the migration, and the code that has the potential to recreate that `Role` will also be gone. How does that sound?

**Comment by biyeun** [comment]
offlined, but just noting here that I would like to see some tests for `revoke_grants` ^_^

**Comment by gherceg** [comment]
@biyeun added some tests, which made me rename some things to be a bit more clear to me (and hopefully others). I think ideally I would have reworked this to raise exceptions or something rather than patching the logger to verify some conditions, but it works.

**Review by biyeun — APPROVED** [review]
with the tests, this looks good to go now. definitely make sure to do the dry run though (dry run functionality/safe guards look clear to me too)

**Pull request merged**